### PR TITLE
Enable Dynamic Memory Budget Solver

### PR DESCRIPTION
**Pull request by basilwong**
Summary:
Full Context: https://docs.google.com/document/d/1-j5KSbfGFJQcH4sYh7BIeJXso3zYzl5G5yFQqXdKx_o/edit?usp=sharing

tl;dr

This change introduces classes which help determine a dynamic memory budget. This will mostly be helpful for models with many implicit graph breaks.

---

New Classes:

*GraphInfoProvider*
* Takes the joint_graph as well as the input memories and runtimes and parses the graph + values into usable forms for the SolverEvaluator.

*KnapsackEvaluator*
* Provides a function: Given all of the four inputs (solver function as a callable, max_dynamic_memory_budget, min_dynamic_memory_budget, dynamic_memory_budget_pareto_granularity) it returns an approximation of the knee point of the pareto distribution.

Test Plan:
### Local E2E Test

https://www.internalfb.com/mlhub/pipeline/1116570856577237
### Distributed E2E Test

aps-fb_fm_v4_768_01_dynamic_updated-b4db74faa6

Differential Revision: D67549590

### CODE REVIEW

File: torch/_functorch/partitioners.py [hunk 1]
```diff
@@ -1393,7 +1393,6 @@ def _optimize_runtime_with_given_memory(
     memory: List[float],
     runtimes: List[float],
     max_memory: float,
-    node_info: NodeInfo,
```
Comment (basilwong) [inline torch/_functorch/partitioners.py:1396]: This is not used since all the info required is in the joint graph.

File: torch/_functorch/partitioners.py [hunk 2]
```diff
@@ -1404,10 +1403,11 @@ def _optimize_runtime_with_given_memory(
     elif SOLVER == "dp":
         return dp_knapsack(memory, runtimes, max_memory)
     elif callable(SOLVER):
-        saved_node_idx, recomp_node_idx = SOLVER(
-            memory, joint_graph, max_memory, node_info, all_recomputable_banned_nodes
+        solver_instance = SOLVER()
```
Comment (basilwong) [inline torch/_functorch/partitioners.py:1406]: Updating how the callable should be defined. Here is an example:

```
class ExampleCustomKnapsackSolver:
    def __init__(self) -> None:
        pass

    def __call__(
        self,
        joint_graph: torch.fx.Graph,
        all_recomputable_banned_nodes: List[torch.fx.Node],
        memory: List[float],
        runtimes: List[float],
        max_memory: float,
    ) -> Tuple[float, List[int], List[int]]:
      ...
      return expected_runtime, saved_node_idx, recomp_node_idx
```

Comment (jansel) [reply torch/_functorch/partitioners.py:1406]: This API seems unnessarily complex.  You should be able to add a wrapper and still use the old API:
```py
def call_solver(memory, joint_graph, max_memory, node_info, all_recomputable_banned_nodes)
   tmp = ExampleCustomKnapsackSolver()
   return tmp(memory, joint_graph, max_memory, node_info, all_recomputable_banned_nodes)

SOLVER = call_solver
```

Though adding the extra `expected_runtime` return which seems fine.

**Review by jansel — CHANGES_REQUESTED** [review]
